include/finalfusion-cxx/Embeddings.hh

```diff
@@ -22,6 +22,55 @@ public:
    */
   Embeddings(std::string const &filename, bool mmap);
 
+  /**
+   * Method to construct Embeddings from a fastText file.
+   *
+   * @param filename path to embeddings.
+   * @throws runtime_error if Embeddings could not be read.
+   */
+  static Embeddings *read_fasttext(std::string const &filename);
+
+  /**
+   * Method to construct Embeddings from a text file.
+   *
+   * Read the word embeddings from a text stream. The text should contain one
+   * word embedding per line in the following format:
+   *
+   * *word0 component_1 component_2 ... component_n*
+   *
+   * @param filename path to embeddings.
+   * @throws runtime_error if Embeddings could not be read.
+   */
+  static Embeddings *read_text(std::string const &filename);
+
+  /**
+   * Method to construct Embeddings from a text file with dimensions.
+   *
+   * Read the word embeddings from a text stream. The text must contain
+   * as the first line the shape of the embedding matrix:
+   *
+   * *vocab_size n_components*
+   *
+   * The remainder of the stream should contain one word embedding per line in
+   * the following format:
+   *
+   * *word0 component_1 component_2 ... component_n*
+   *
+   * @param filename path to embeddings.
+   * @throws runtime_error if Embeddings could not be read.
+   */
+  static Embeddings *read_text_dims(std::string const &filename);
+
+  /**
+   * Method to construct Embeddings from a word2vec binary file.
+   *
+   * Read the word embeddings from a file in word2vec binary format.
+   *
+   * @param filename path to embeddings.
+   * @throws runtime_error if Embeddings could not be read.
+   */
+  static Embeddings *read_word2vec(std::string const &filename);
```

Comment:
Almost perfect, could these be changed to return `Embeddings` rather than `Embeddings *`?

Reply:
I tried to get that to work but e.g.:

```C++
Embeddings Embeddings::read_fasttext(std::string const &filename) {
  return Embeddings(new EmbeddingsImpl(filename, NonFifuFormat::FastText));
};
```
doesn't compile because:
```
/data/finalfusion-cxx/src/Embeddings.cc: In static member function ‘static Embeddings Embeddings::read_fasttext(const string&)’:
/data/finalfusion-cxx/src/Embeddings.cc:78:74: error: use of deleted function ‘Embeddings::Embeddings(const Embeddings&)’
   return Embeddings(new EmbeddingsImpl(filename, NonFifuFormat::FastText));

In file included from /data/finalfusion-cxx/src/Embeddings.cc:5:0:
/data/finalfusion-cxx/include/finalfusion-cxx/Embeddings.hh:14:7: note: ‘Embeddings::Embeddings(const Embeddings&)’ is implicitly deleted because the default definition would be ill-formed:
 class Embeddings {
       ^~~~~~~~~~
/data/finalfusion-cxx/include/finalfusion-cxx/Embeddings.hh:14:7: error: use of deleted function ‘std::unique_ptr<_Tp, _Dp>::unique_ptr(const std::unique_ptr<_Tp, _Dp>&) [with _Tp = EmbeddingsImpl; _Dp = std::default_delete<EmbeddingsImpl>]’
In file included from /usr/include/c++/7/memory:80:0,
                 from /data/finalfusion-cxx/src/Embeddings.cc:2:
/usr/include/c++/7/bits/unique_ptr.h:388:7: note: declared here
       unique_ptr(const unique_ptr&) = delete;
       ^~~~~~~~~~
```

which I didn't manage to work through. Do you have an idea how this would work?

Reply:
I guess it requires copy construction. I forgot the exact visibility rules, what if you make the copy constructor private rather than deleting it.

Otherwise, I guess it's time to think about implementing copy construction / assignment.

Reply:
> I guess it requires copy construction. I forgot the exact visibility rules, what if you make the copy constructor private rather than deleting it.

Although, I didn't explicitly delete that, adding a private copy constructor fixes compilation of the library. The testcase with 
```C++
std::unique_ptr<Embeddings> embeddings = std::unique_ptr<Embeddings>(Embeddings::read_word2vec("testdata/test.w2v"));
```
fails to compile on the other hand since `read_word2vec` now returns `Embeddings' which can't be used in the `unique_ptr` constructor. This could of course be fixed by declaring `embeddings` as `Embeddings` instead of using a unique pointer and leaving `embeddings` on the stack, but then we'd run into issues with `finalfusion-tensorflow`, where `Embeddings` is wrapped in a `unique_ptr`.

Is there a way to use the factory methods to store objects on the heap?

Reply:
> ```c++
> std::unique_ptr<Embeddings> embeddings = std::unique_ptr<Embeddings>(Embeddings::read_word2vec("testdata/test.w2v"));
> ```
> 
> fails to compile on the other hand since `read_word2vec` now returns `Embeddings`which can't be used in the `unique_ptr` constructor. This could of course be fixed by declaring `embeddings` as `Embeddings` instead of using a unique pointer

use `make_unique` or if that is not available (because C++ < 14), I guess you could copy construct it into unique. Something like:

```
std::unique<Embeddings>(new Embeddings(Embeddings::read_word2vec("testdata/test.w2v")))
```

But my biggest qualm with the factories was returning raw pointers. So, if you feel it's more ergonomical, you could also let the factory methods return `unique_ptr`s.

Reply:
> ```c++
> Embeddings embeds = Embeddings::read_word2vec("testdata/test.w2v");
> std::unique_ptr<Embeddings> embeddings = std::unique_ptr<Embeddings>(&embeds);
> ```
> 
> fixes the testcase but only after making the copy constructor public.

BTW, I think this would actually be invalid, `embeds` is owned by the stack, so this would result in `std::unique_ptr` trying to deallocate stack memory later.

Reply:
> `std::unique<Embeddings>(new Embeddings(Embeddings::read_word2vec("testdata/test.w2v")))`

This only works with a public copy constructor, is there a clear downside to this?

> But my biggest qualm with the factories was returning raw pointers. So, if you feel it's more ergonomical, you could also let the factory methods return unique_ptrs.

My thought was that returning raw pointers would most closely resemble using a standard constructor, i.e. `new Embeddings` also returns a raw pointer.

Reply:
> This only works with a public copy constructor, is there a clear downside to this?

Not really, as long as you are copying the members properly or wrap them such that they can be shared between instances (`shared_ptr`).

Copy construction had a bad name, because C++ would implicitly create a copy constructor. Then when people had raw pointer members, this would result in wild pointers if one of the instances goes out of scope (and `delete`s the pointer). But if you implement copy construction and assignment properly, it's actually nice to have.

Reply:
I'm still a bit lost here. Would you suggest a `shared_ptr` to the `EmbeddingsImpl` rather than `unique_ptr`? Else I'm not sure about how to copy that `unique_ptr`'s data to initialize a new object.

As of now, I just have a publicly declared copy constructor in the header without providing an implementation myself. I'd assume that can lead to unwanted behaviour since there could be several `unique_ptr`s to the same data?

Reply:
> I'm still a bit lost here. Would you suggest a `shared_ptr` to the `EmbeddingsImpl` rather than `unique_ptr`?

Yes, the refcounting in `shared_ptr` is probably negligible compared to looking up embeddings.

> As of now, I just have a publicly declared copy constructor in the header without providing an implementation myself. I'd assume that can lead to unwanted behaviour since there could be several `unique_ptr`s to the same data?

AFAIR `unique_ptr` has a move constructor, so the old (non-copied) unique pointer will probably be reset to the null pointer, which would be bad ;). But you'll have to check the documentation, because this stuff was added & supported after I stopped using C++.

Reply:
Thanks for the explanation! I changed `embeddings_impl_` to be a `shared_ptr` now.

> AFAIR `unique_ptr` has a move constructor, so the old (non-copied) unique pointer will probably be reset to the null pointer, which would be bad ;). But you'll have to check the documentation, because this stuff was supported in C++ compilers after I stopped using C++ (we only had `auto_ptr` plus some pointer classes in Boost, no move constructors yet).
> 
> I still have Scott Meyer's book on all the >= C++11 stuff on the bookshelf looking at me. But after reading a few pages when I bought it, I decided that C++ is now even are more complex beast that I don't want to bother with.

Still beats my experience with C++ by quite a large margin ;) My knowledge is restricted to what Google & SO revealed to me while looking for specific problems. Working with C++ definitely makes me appreciate Rust, the hassle to get a valid program with C++ is definitely bigger than dealing with the Rust compiler.